src/dmd/root/array.d

```diff
@@ -591,3 +591,68 @@ unittest
     static assert (!is(typeof(Array!(OtherStruct).init.sort())));
     static assert (!is(typeof(Array!(OtherStruct).init.sort!pred)));
 }
+
+/**
+ * Filters the given array based on the given predicate.
+ *
+ * Returns: a new array containing only elements for which the predicate returns
+ *  `true`
+ */
+inout(T)[] filter(alias predicate, T)(inout T[] array)
```

Comment:
I've seen advice on naming these functions `filtered` and `mapped` instead to make it clear that we are not returning a lazy range.

Reply:
I disagree. Phobos contain some weird naming conventions just because initially there were some eager implementations, then someone thought it would be a good idea to add lazy implementations as well. The existing implementations could not be removed because that would breaking existing code. So new names had to be invented.

I instead prefer to look at the return type, which tells the truth.

Reply:
> I instead prefer to look at the return type, which tells the truth.

True, but then you can't overload the the lazy versions using the same names `filter` and `map`, right?

Would you have like Phobos to have different namings for lazy vs eager `map` and `filter`? If so what would you like that naming to look like?

Reply:
> True, but then you can't overload the the lazy versions using the same names filter and map, right?

No.

> Would you have like Phobos to have different namings for lazy vs eager map and filter? If so what would you like that naming to look like?

Ideally no. But Phobos only has lazy versions.

Reply:
> > True, but then you can't overload the the lazy versions using the same names filter and map, right?
> 
> No.

How do you overload a lazy version on arrays?

Reply:
Sorry, I meant: no you cannot overload lazy and non-lazy versions using the same names.